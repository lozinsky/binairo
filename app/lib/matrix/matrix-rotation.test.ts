import { expect, test } from 'vitest';

import {
  DoubleMatrixRotation,
  type MatrixReversible,
  type MatrixRotatable,
  QuadrupleMatrixRotation,
} from './matrix-rotation';

test('double rotates', () => {
  class Target implements MatrixRotatable<Target> {
    readonly name: string;

    constructor(name: string) {
      this.name = name;
    }

    rotate() {
      return new Target(`rotated ${this.name}`);
    }
  }

  const rotation = new DoubleMatrixRotation(new Target('target'));

  expect(rotation.rotate(0)).toMatchSnapshot();
  expect(rotation.rotate(1)).toMatchSnapshot();
  expect(rotation.rotate(2)).toMatchSnapshot();
});

test('quadruple rotates', () => {
  class Target implements MatrixRotatable<Target>, MatrixReversible<Target> {
    readonly name: string;

    constructor(name: string) {
      this.name = name;
    }

    reverse() {
      return new Target(`reversed ${this.name}`);
    }

    rotate() {
      return new Target(`rotated ${this.name}`);
    }
  }

  const rotation = new QuadrupleMatrixRotation(new Target('target'));

  expect(rotation.rotate(0)).toMatchSnapshot();
  expect(rotation.rotate(1)).toMatchSnapshot();
  expect(rotation.rotate(2)).toMatchSnapshot();
  expect(rotation.rotate(3)).toMatchSnapshot();
  expect(rotation.rotate(4)).toMatchSnapshot();
});
