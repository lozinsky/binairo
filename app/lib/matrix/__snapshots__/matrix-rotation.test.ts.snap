// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`double rotates 1`] = `
Target {
  "name": "target",
}
`;

exports[`double rotates 2`] = `
Target {
  "name": "rotated target",
}
`;

exports[`double rotates 3`] = `null`;

exports[`quadruple rotates 1`] = `
Target {
  "name": "target",
}
`;

exports[`quadruple rotates 2`] = `
Target {
  "name": "reversed target",
}
`;

exports[`quadruple rotates 3`] = `
Target {
  "name": "rotated target",
}
`;

exports[`quadruple rotates 4`] = `
Target {
  "name": "reversed rotated target",
}
`;

exports[`quadruple rotates 5`] = `null`;
